futures/src/channel.rs

```diff
@@ -145,73 +145,104 @@ pub struct ChannelFlowOptions {
 
 impl<T: AsyncRead+AsyncWrite+Send+'static> Channel<T> {
     /// create a channel
-    pub fn create(transport: Arc<Mutex<AMQPTransport<T>>>) -> Box<Future<Item = Self, Error = io::Error> + Send + 'static> {
+    pub fn create(transport: Arc<Mutex<AMQPTransport<T>>>) -> impl Future<Item = Self, Error = io::Error> + Send {
         let channel_transport = transport.clone();
         let create_channel = future::poll_fn(move || {
-            let mut transport = try_lock_transport!(channel_transport);
+            let mut transport = match channel_transport.try_lock() {
+                Ok(t) => t,
+                Err(_) => if channel_transport.is_poisoned() {
+                    return Err(io::Error::new(io::ErrorKind::Other, "Transport mutex is poisoned"));
+                } else {
+                    task::current().notify();
+                    return Ok(Async::NotReady);
+                }
+            };
             return Ok(Async::Ready(Channel {
                 id:        transport.conn.create_channel(),
                 transport: channel_transport.clone(),
             }))
         });
 
-        Box::new(create_channel.and_then(|channel| {
+        create_channel.and_then(|channel| {
             let channel_id = channel.id;
             channel.run_on_locked_transport("create", "Could not create channel", move |transport| {
                 transport.conn.channel_open(channel_id, "".to_string()).map(Some)
             }).map(move |_| {
                 channel
             })
-        }))
+        })
     }
 
     /// request access
     ///
     /// returns a future that resolves once the access is granted
-    pub fn access_request(&self, realm: &str, options: &AccessRequestOptions) -> Box<Future<Item = (), Error = io::Error> + Send + 'static> {
-        self.run_on_locked_transport("access_request", "Could not request access", |transport| {
-            transport.conn.access_request(self.id, realm.to_string(),
+    pub fn access_request(&self, realm: &str, options: &AccessRequestOptions) -> impl Future<Item = (), Error = io::Error> + Send {
```

Comment:
Why the move and all the cloning everywhere which wasn't needed before?

Reply:
The other my option for `impl Trait` result was the following signature:
```rust
    pub fn access_request<'a>(&self, realm: &str, options: &'a AccessRequestOptions) -> impl Future<Item = (), Error = io::Error> + Send + 'a {
```

Here I decided to just clone `options` parameter and put less restrictions for return type. The cloning should be cheap and anyways there is `.to_string()` invocation which is more expensive because of allocation.

Reply:
But why is that needed now and wasn't before? Because of the FnOnce change?